### PR TITLE
Convert Contribution zipfile to Pack

### PR DESCRIPTION
**Pull request by avidan-H**
## Status
- [x] In Progress

## Related Issues
fixes: https://github.com/demisto/etc/issues/26061

## Description
Takes a contribution zip downloaded from Demisto and converts the contents to the proper Pack structure.

## Screenshots
**zip file and its unzipped contents**  

---

zipfile_and_unzipped_contents

**created pack after running the command**

---

image

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: demisto_sdk/commands/init/README.md [hunk 1]
```diff
@@ -42,3 +46,8 @@ once given a script will be created under "Scripts" directory and the yml file w
 
 `demisto-sdk init --pack -n My_Pack`
 This will create a new pack named "My_Pack" under the "Packs" directory in content repo.
+
+`demisto-sdk init --contribution ~/Downloads/my_contribution.zip`
```
Comment (Itay4) [inline demisto_sdk/commands/init/README.md:50]: is there a limitation that this need to be ran from the content repo root?

File: demisto_sdk/commands/init/initiator.py [hunk 2]
```diff
@@ -46,7 +51,8 @@ class Initiator:
     """
 
     def __init__(self, output: str, name: str = '', id: str = '', integration: bool = False, script: bool = False,
-                 pack: bool = False, demisto_mock: bool = False, common_server: bool = False):
+                 pack: bool = False, demisto_mock: bool = False, common_server: bool = False,
+                 contribution: Union[str, None] = None):
```
Comment (Itay4) [inline demisto_sdk/commands/init/initiator.py:55]:
```suggestion
                 contribution: Optional[str] = None):
```

File: demisto_sdk/commands/init/initiator.py [hunk 3]
```diff
@@ -99,6 +108,74 @@ def init(self):
             self.get_created_dir_name(created_object="pack")
             return self.pack_init()
 
+    def convert_contribution_to_pack(self):
+        '''Create a Pack in the content repo from the contents of a contribution zipfile'''
```
Comment (Itay4) [inline demisto_sdk/commands/init/initiator.py:112]:
```suggestion
        """Create a Pack in the content repo from the contents of a contribution zipfile"""
```

File: demisto_sdk/commands/init/initiator.py [hunk 4]
```diff
@@ -180,12 +250,28 @@ def pack_init(self) -> bool:
 
         return True
 
+    def create_pack_base_files(self):
+        '''
```
Comment (Itay4) [inline demisto_sdk/commands/init/initiator.py:254]:
```suggestion
        """
```

File: demisto_sdk/commands/init/initiator.py [hunk 5]
```diff
@@ -180,12 +250,28 @@ def pack_init(self) -> bool:
...
+        Create empty 'README.md', '.secrets-ignore', and '.pack-ignore' files that are expected
+        to be in the base directory of a pack
+        '''
```
Comment (Itay4) [inline demisto_sdk/commands/init/initiator.py:257]:
```suggestion
        """
```

File: demisto_sdk/commands/init/initiator.py [hunk 6]
```diff
@@ -180,12 +250,28 @@ def pack_init(self) -> bool:
...
+
+        fp = open(os.path.join(self.full_output_path, '.secrets-ignore'), 'a')
+        fp.close()
+
+        fp = open(os.path.join(self.full_output_path, '.pack-ignore'), 'a')
+        fp.close()
+
     @staticmethod
-    def create_metadata(fill_manually: bool) -> Dict:
+    def create_metadata(fill_manually: bool, data: dict = {}) -> Dict:
```
Comment (Itay4) [inline demisto_sdk/commands/init/initiator.py:269]:
```suggestion
    def create_metadata(fill_manually: bool, data: Dict = {}) -> Dict:
```

File: demisto_sdk/commands/init/initiator.py [hunk 7]
```diff
@@ -99,6 +108,74 @@ def init(self):
             self.get_created_dir_name(created_object="pack")
             return self.pack_init()
 
+    def convert_contribution_to_pack(self):
+        '''Create a Pack in the content repo from the contents of a contribution zipfile'''
+        packs_dir = os.path.join(get_content_path(), 'Packs')
+        metadata_dict = {}
+        with zipfile.ZipFile(self.contribution) as zipped_contrib:
+            with zipped_contrib.open('metadata.json') as metadata_file:
+                print_color(f'Pulling relevant information from {metadata_file.name}', LOG_COLORS.NATIVE)
```
Comment (Itay4) [inline demisto_sdk/commands/init/initiator.py:117]: lets use `click.echo()` instead of `print_color`

File: demisto_sdk/commands/init/initiator.py [hunk 8]
```diff
@@ -99,6 +108,74 @@ def init(self):
             self.get_created_dir_name(created_object="pack")
             return self.pack_init()
 
+    def convert_contribution_to_pack(self):
+        '''Create a Pack in the content repo from the contents of a contribution zipfile'''
+        packs_dir = os.path.join(get_content_path(), 'Packs')
+        metadata_dict = {}
+        with zipfile.ZipFile(self.contribution) as zipped_contrib:
```
Comment (Itay4) [inline demisto_sdk/commands/init/initiator.py:115]: shouldnt we wrap this whole thing in `try-except` ?

File: demisto_sdk/commands/init/initiator.py [hunk 9]
```diff
@@ -99,6 +108,74 @@ def init(self):
...
+        self.full_output_path = pack_dir
+        self.create_pack_base_files()
+        metadata_dict = Initiator.create_metadata(fill_manually=False, data=metadata_dict)
+        metadata_path = os.path.join(self.full_output_path, 'pack_metadata.json')
+        with open(metadata_path, 'w') as pack_metadata_file:
+            json.dump(metadata_dict, pack_metadata_file, indent=4)
+        # remove metadata.json file
+        os.remove(os.path.join(pack_dir, 'metadata.json'))
+
+    def content_item_to_package_format(self, content_item_dir, del_unified):
```
Comment (Itay4) [inline demisto_sdk/commands/init/initiator.py:158]: add type hints here

File: demisto_sdk/commands/init/initiator.py [hunk 10]
```diff
@@ -99,6 +108,74 @@ def init(self):
...
+                    self.content_item_to_package_format(content_item_dir, del_unified=True)
+        # create pack's base files
+        self.full_output_path = pack_dir
+        self.create_pack_base_files()
+        metadata_dict = Initiator.create_metadata(fill_manually=False, data=metadata_dict)
+        metadata_path = os.path.join(self.full_output_path, 'pack_metadata.json')
+        with open(metadata_path, 'w') as pack_metadata_file:
+            json.dump(metadata_dict, pack_metadata_file, indent=4)
+        # remove metadata.json file
+        os.remove(os.path.join(pack_dir, 'metadata.json'))
```
Comment (Itay4) [inline demisto_sdk/commands/init/initiator.py:156]: where is this file created ?

File: demisto_sdk/commands/init/initiator.py [hunk 11]
```diff
@@ -99,6 +108,74 @@ def init(self):
...
+        shutil.unpack_archive(filename=self.contribution, extract_dir=pack_dir)
+        pack_subdirectories = get_child_directories(pack_dir)
+        for pack_subdir in pack_subdirectories:
+            basename = os.path.basename(pack_subdir)
+            if basename in ENTITY_TYPE_TO_DIR:
+                dst_name = ENTITY_TYPE_TO_DIR.get(basename)
+                src_path = os.path.join(pack_dir, basename)
+                dst_path = os.path.join(pack_dir, dst_name)
+                content_item_dir = shutil.move(src_path, dst_path)
+                if basename in {'integration', 'automation'}:
```
Comment (Itay4) [inline demisto_sdk/commands/init/initiator.py:146]: lets use the constants we have for these entities
lets also search for `script` in case we add it in the future

File: demisto_sdk/commands/init/initiator.py (hunk 9 above)
Comment (Itay4) [inline demisto_sdk/commands/init/initiator.py:158]: lets set the default value of `del_unified` to be `True`

File: demisto_sdk/commands/init/initiator.py [hunk 12]
```diff
@@ -99,6 +108,74 @@ def init(self):
...
+
+        Args:
+            content_item_dir (str): Path to the directory containing the content item YAML file(s)
+            del_unified (bool): Whether to delete the unified yaml the package was extracted from
+        '''
+        child_files = get_child_files(content_item_dir)
+        content_item_file_path = ''
+        for child_file in child_files:
+            cf_name_lower = os.path.basename(child_file).lower()
+            if cf_name_lower.startswith(('integration', 'automation')) and cf_name_lower.endswith('yml'):
```
Comment (Itay4) [inline demisto_sdk/commands/init/initiator.py:171]: lets use the constants we have for these entities
lets also search for `script` in case we add it in the future

**Comment by Itay4** [comment]
not related, but i see we have 2 unused constants, lets remove them: `PACK_TEMPLATE_FOLDER` and `PACK_METADATA_TEMPLATE`

File: .pre-commit-config.yaml [hunk 13]
```diff
@@ -22,10 +22,19 @@ repos:
...
--   repo: https://github.com/pre-commit/mirrors-isort
-    rev: v4.3.21
+-   repo: https://github.com/timothycrosley/isort
+    rev: 4.3.21-2
     hooks:
     -   id: isort
+        args: [--diff]
+        types: [file, python, text]
+        verbose: true
+-   repo: https://github.com/timothycrosley/isort
```
[the remark this answers is not in the document]
Comment (avidan-H) [reply .pre-commit-config.yaml:32]: The mirror we were using for the `isort` hook said it was deprecated. It also really bothered me that when `isort` would fail, you wouldn't actually see the modification that was made (and ergo the reason why your build failed on `isort`).

File: demisto_sdk/commands/init/README.md (hunk 1 above)
Comment (avidan-H) [reply demisto_sdk/commands/init/README.md:50]: It needs to be run from within the content repo but it doesn't matter from where in the repo.

File: demisto_sdk/commands/init/initiator.py (hunk 10 above)
Comment (avidan-H) [reply demisto_sdk/commands/init/initiator.py:156]: It isn't. It comes in the zip file downloaded from server.

**Comment by Itay4** [comment]
@jochman please review changes in test utils

File: TestSuite/contribution.py [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+        self.create_metadata_for_zip()
+        if zip_dst:
+            self.created_zip_filepath = shutil.make_archive(
+                zip_dst / self.name, 'zip', self.target_dir
+            )
+        else:
+            self.created_zip_filepath = shutil.make_archive(
+                self.target_dir.parent / self.name, 'zip', self.target_dir
+            )
+        if del_src_files:
```
Comment (jochman) [inline TestSuite/contribution.py:234]: why is this needed?

Comment (avidan-H) [reply TestSuite/contribution.py:234]: I don't want the source files polluting the temporary repo I'm the test is running in.

**Comment by avidan-H** [comment]
> It looks like you can inherit from the integration object or just add the `create_zip` function to integration.
> Why you've decided to duplicate it?

`Contribution` is actually more similar to the `Pack` class but that being said, it still has lots of differences.

File: TestSuite/integration.py [hunk 15]
```diff
@@ -1,16 +1,18 @@
...
+from demisto_sdk.commands.unify.unifier import Unifier
 from deprecated import deprecated
 from TestSuite.file import File
 from TestSuite.test_tools import suite_join_path
 from TestSuite.yml import YAML
 
 
 class Integration:
-    def __init__(self, tmpdir: Path, name, repo):
+    def __init__(self, tmpdir: Path, name, repo, unified: bool = False):
```
Comment (Itay4) [inline TestSuite/integration.py:15]:
```suggestion
    def __init__(self, tmpdir: Path, name, repo, create_unified: bool = False):
```

File: TestSuite/integration.py [hunk 16]
```diff
@@ -20,6 +22,9 @@ def __init__(self, tmpdir: Path, name, repo):
         self._tmpdir_integration_path = tmpdir / f'{self.name}'
         self._tmpdir_integration_path.mkdir()
 
+        # if creating a unified yaml
+        self.create_unified = unified
```
Comment (Itay4) [inline TestSuite/integration.py:26]:
```suggestion
        self.create_unified = create_unified
```

File: TestSuite/script.py [hunk 17]
```diff
@@ -1,14 +1,17 @@
...
+from demisto_sdk.commands.unify.unifier import Unifier
 from TestSuite.integration import Integration
 from TestSuite.test_tools import suite_join_path
 
 
 class Script(Integration):
     # Im here just to have one!!!
-    def __init__(self, tmpdir: Path, name, repo):
-        super().__init__(tmpdir, name, repo)
+    def __init__(self, tmpdir: Path, name, repo, unified=False):
```
Comment (Itay4) [inline TestSuite/script.py:13]:
```suggestion
    def __init__(self, tmpdir: Path, name, repo, create_unified =False):
```

File: TestSuite/script.py [hunk 18]
```diff
@@ -1,14 +1,17 @@
...
+        super().__init__(tmpdir, name, repo, unified)
```
Comment (Itay4) [inline TestSuite/script.py:14]:
```suggestion
        super().__init__(tmpdir, name, repo, create_unified)
```

File: demisto_sdk/commands/init/tests/initiator_test.py [hunk 19]
```diff
@@ -181,3 +184,75 @@ def test_yml_reformatting(tmp_path, initiator):
...
+    - The zipfile contains a unified script file
+    - The zipfile contains a unified integration file
+    When
+    - Converting the zipfile to a valid Pack structure
+    Then
+    - Ensure script and integration are componentized and in valid directory structure
+    '''
+    # Create all Necessary Temporary directories
+    # create temp directory for the repo
+    repo_dir = tmp_path / 'content_repo'
```
Comment (Itay4) [inline demisto_sdk/commands/init/tests/initiator_test.py:214]: i think it would make more sense to use here `tmpdir` fixture

**Comment by Itay4** [comment]
@avidan-H is there value for an integration test here as well?

**Comment by avidan-H** [comment]
> @avidan-H is there value for an integration test here as well?

No, I don't think so. The unit test I added was more akin to an integration test than a unit test tbh.

File: demisto_sdk/commands/init/tests/initiator_test.py (hunk 19 above)
Comment (avidan-H) [reply demisto_sdk/commands/init/tests/initiator_test.py:214]: Unless there is an objection, I'm going to leave it as-is for now, since changing it to use the `tmpdir` fixture would require some reworking of the unit test.

File: TestSuite/contribution.py [hunk 20]
```diff
@@ -0,0 +1,235 @@
+import shutil
+from pathlib import Path
+from typing import List, Optional
+
+from TestSuite.integration import Integration
+from TestSuite.json_based import JSONBased
+from TestSuite.script import Script
+# from TestSuite.secrets import Secrets
```
Comment (Itay4) [inline TestSuite/contribution.py:8]: remove comment
